### PR TITLE
Add initial design for OperatorPolicy

### PR DESCRIPTION
**Pull request by JustinKuli**
Refs:
 - https://issues.redhat.com/browse/ACM-3097

For easier reading: https://github.com/JustinKuli/ocm-enhancements/blob/89-operator-policy/enhancements/sig-policy/89-operator-policy-kind/README.md

### CODE REVIEW

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (gparvin) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:39]: Today there are cases where I think something happens in the wrong order and you end up with a failed operator install.  You can't resolve except by doing some cleanup and then trying again.  Does that sound familiar and would that case be covered because the "ordering" is no longer a concern?  Or would there still be a need for a retry on failure?

Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:39]: By having the new controller manage the "imperative" steps that could be done in the wrong order, I think it can avoid at least some of the issues we've seen before when we just apply all the resources at once with ConfigurationPolicies.

It might be impossible to prevent all installation errors, but depending on what's required to "cleanup and try again", I think that could also be handled by the controller automatically.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (fgiloux) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:18]: > The new controller would better handle imperative installation steps for OLM objects

I guess it is referring to the fact that you need to approve the installation when the install plan approval strategy has been set to manual in the subscription.
I would propose to rephrase to: "The new controller would better handle installation requiring approvals of OLM operators."
Technically it is possible to handle everything in a declarative way. At the end of the day the API is a resource, whose fields can be modified. I understand the challenge with InstallPlans that are a second level resources (not directly created) but need to be modified for the approval.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (fgiloux) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:27]: > Upgrading or uninstalling an operator can also be difficult with ConfigurationPolicies, since it is not generally as easy as just updating one value or deleting a resource.

Can you be specific? It should actually be the case for upgrades. Is it because of the InstallPlan approval?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (fgiloux) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:106]: > The spec.requireLatestVersion field configures whether the "latest" version of the operator defined in the catalog on this cluster "should" be installed.

Wouldn't it be the default behavior?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (fgiloux) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:279]: I think so.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (jc-rh) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:83]: What does none mean?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (jc-rh) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:27]: In addition to the need of install plan approval, issues we ran into: (may be specific to disconnected env)
- catalog src policy becomes compliant before it's truly ready
- subscription policy with 'AtLatestKnown` takes a while to become non-compliant after catalog src is ready
- after install plan/subscription status is updated, there is still a chance for the csv to be stuck in failed state

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (mprahl) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:63]: What is checked in `inform` mode? What kind of non-compliant message would you get?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (mprahl) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:110]: Is this the default?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (mprahl) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:242]: Did you mean this?
```suggestion
      status: 'False'
      type: Compliant
```

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (mprahl) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:279]: I like this idea.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:110]: It probably should be... I'm still thinking about the best way to do that. Since I want it to be possible to configure nothing to be deleted, an empty list does not imply the "default" behavior. And filling it with an empty list, vs not specifying the field at all, feels too subtle of a difference - I think it could be confusing.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:1]: Great comments! I think most are covered in my recent update, but I will call out these specific points:

5. I updated the goal to specify that *creation* of CatalogSources will not be covered - as noted by other details, the CatalogSource's health will be monitoried.

7. Renamed `requireLatestVersion` to `latestVersionNecessary`.

9. Added. Initially I wanted to try and keep things smaller here, but that doesn't seem necessary.

10. I didn't want to over-specify in this design what status might be reported. I do agree that what you've mentioned here is important and likely should be reported. I've added an open question for how much specification is required here as part of the design.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:123]: I don't know what you mean.

One case I was trying to account for was where the policy was created with the (current) "newest" version available and tested, but not wanting to immediately update whenever something new comes out. Instead, the user could use this so that they are notified (via a NonCompliant policy) when an update is available, so they can test it out.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:167]: In my opinion of what a user story should be, it doesn't make sense to pull the health reporting out into a separate story. I do agree it could help feature development, but I didn't want to artificially split it up here.

**Comment by brian-jarvis** [comment]
One thing to note, InstallPlans contain details of all operators in the same namespace.  You can't approve/upgrade one individually.   This causes issues for customers today (perhaps olm 2.0 will correct this) trying to manage manually approving operators through cluster environments.

Given the InstallPlan below:
- which OperatorPolicy should own updating it?
- what happens if one OperatorPolicy has limited openshift-gitops to 1.7.x and servicemesh to 2.3.x?  

`apiVersion: operators.coreos.com/v1alpha1
kind: InstallPlan
metadata:
  creationTimestamp: "2023-03-24T15:23:01Z"
  generateName: install-
  generation: 1
  labels:
    operators.coreos.com/kiali-ossm.openshift-operators: ""
    operators.coreos.com/servicemeshoperator.openshift-operators: ""
  name: install-j9k5k
  namespace: openshift-operators
  ownerReferences:
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: devspaces
    uid: 1266cac3-5703-405d-8088-4c5f5cfbee28
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: devworkspace-operator-fast-redhat-operators-openshift-marketplace
    uid: 09cfe1be-bd9c-4ca5-b6ae-26bc101a134f
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: jaeger-product
    uid: b58625e0-2e4a-4229-8dcf-b73d28e411ef
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: kiali-ossm
    uid: 7023b9d3-6c48-4611-951f-e36d72289350
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: openshift-gitops-operator
    uid: 33f2927b-6033-42eb-b743-a1606fb35d4f
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: openshift-pipeline-operator
    uid: 68200532-924d-46e0-8882-5bf3b0478e6c
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: rh-service-binding-operator
    uid: 0e093920-e581-4fde-8438-18a33ff1bb88
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: Subscription
    name: servicemeshoperator
    uid: ec09113a-7aac-46f9-be44-37ffa55fc79a
  resourceVersion: "1891393943"
  uid: 5b4c51ca-a97c-40a2-8cae-f66e4c6e8d11
spec:
  approval: Automatic
  approved: true
  clusterServiceVersionNames:
  - openshift-gitops-operator.v1.8.1
  - kiali-operator.v1.57.6
  - servicemeshoperator.v2.3.2
  - openshift-pipelines-operator-rh.v1.9.2
`

**Comment by JustinKuli** [comment]
Newest revision notes. Overall the idea was to improve ease-of-use, and incorporate comments from the previous reviews.

The configuration of what upgrades are allowed has been changed from the `endingCSV` field that was in `spec.subscription`, to a new field `spec.allowedCSVs`. This makes the subscription field almost exactly match the OLM subscription spec (only adding a `namespace` field), and makes the behavior of the policy in `inform` mode more obvious.

The `spec.latestVersionNecessary` field has been brought into a new section `spec.violationConditions` which will hold more configurations of that type: things that *could* be cause for a violation, but not necessarily for all users/policies.

The specification of the `status` is more complete now.

The compliance event now has an example YAML.

There are more User Stories, and they are generally smaller / more specific, ideally each should be implementable in a reasonable amount of time (some requiring other stories to already be completed).

Some details on approving InstallPlans has been added to the notes, with observations from existing InstallPlans on clusters.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (jc-rh) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:139]: Is there a valid case for this to be set with empty value? If not, is it better to keep it optional and allow all if not set?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (irinamihai) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:113]: How would we use this to ensure the Subscription's status has **status.state: AtLatestKnown** without specifying the exact CSV?

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (jc-rh) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:63]: I agree that the catalog source should be handled in a different policy. Just to highlight that currently the catalog source API does not provide `observerdGeneration` in its status. Any thing we derive from its status can be stale if its spec just got updated. This can make the policy ordering ineffective. The operator policy can go/stay compliant when it shouldn't be. Same concern goes with the `catalogSourceUnhealthy` field in `statusConfig`. It wouldn't do its job if we set it to `NonCompliant` in cases like this. I opened an RFE for this but it is set to lowest priority possible ...
https://issues.redhat.com/browse/RFE-3846

I've linked it to this enhancement JIRA. If we can push for it together and get it implemented by OLM, this enhancement will be more successful :)

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (jc-rh) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:151]: I guess if we put `NonCompliant` for both `upgradesAvailable` and `upgradesProgressing`, it might cover 99% of the whole process. It might be hard to guarantee it will always work like that as things evolve though. Is it better to introduce an explicit field for `atLatestKnown`? This is the main use case for telco and it's part of our current du reference spec.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:139]: Good question! I hadn't realized it, but yes, an empty list here could be useful: in particular when the policy is in `inform` mode, that could be used to verify that an operator is *not* installed (similar to `mustnothave` in ConfigurationPolicy).

The design right now doesn't specify whether already-existing subscriptions should be deleted by an OperatorPolicy in `enforce` mode, that might need to be considered...

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:113]: Setting `upgradesAvailable: NonCompliant` would cause the policy to become non-compliant when the Subscription has `status.state: UpgradePending`.

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:151]: I realized I didn't fully specify this, but I was expecting `upgradesAvailable` to basically correspond to the `atLatestKnown` case... is there a situation I'm missing there? I guess I don't know what the Subscription's state will look like if it's not at the newest version, but it can't upgrade to that because of a version incompatibility or something.

Comment (jc-rh) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:151]: Yes, there are other states in addition to these two:
`const (
	SubscriptionStateNone             = ""
	SubscriptionStateFailed           = "UpgradeFailed"
	SubscriptionStateUpgradeAvailable = "UpgradeAvailable"
	SubscriptionStateUpgradePending   = "UpgradePending"
	SubscriptionStateAtLatest         = "AtLatestKnown"
)`

File: enhancements/sig-policy/89-operator-policy-kind/README.md
Comment (serngawy) [inline enhancements/sig-policy/89-operator-policy-kind/README.md:97]: better name could be **approvedCSV** as allowedCSV meaning more like it is allowed to show up for subscriptions but not approved.

Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md:97]: I'm thinking about this in the new possible mode of a "mustnothave" OperatorPolicy. In that case, it would be nice to be able to supply a list of specific versions I *don't* want to allow on the managed cluster.

Would a simpler name like `versions` work? Then in a "musthave" OperatorPolicy, this list indicates the versions that can be present for the policy to be compliant. And in "mustnothave" mode, the list would indicate versions that must not be present.

Comment (serngawy) [reply enhancements/sig-policy/89-operator-policy-kind/README.md]: okay, make sense for mustnothave case.

Comment (jc-rh) [reply enhancements/sig-policy/89-operator-policy-kind/README.md]: Are you saying two policies are needed if we want to allow some versions and disallow some others?

Comment (JustinKuli) [reply enhancements/sig-policy/89-operator-policy-kind/README.md]: @jc-rh it depends on if you know the specific versions you want to allow. If you make a "musthave" policy with specific versions, than any versions other than what is specified will make the policy NonCompliant (and in enforce mode, the controller might try to remove that operator).

If you can't restrict the allowed versions to a specific list - maybe you want something like "any version, except x, y, or z" - then in this proposal, two policies would be necessary: a "musthave" with no versions specified, and a "mustnothave" with x, y, and z. (This case wasn't really covered by any previous design iterations.)

I was anticipating "mustnothave" being more useful if the operators aren't necessarily being managed by an OperatorPolicy.

Comment (jc-rh) [reply enhancements/sig-policy/89-operator-policy-kind/README.md]: Hmm, OK. Our main use case would be `AtLatestKnown`. It would be a pain to update allowed version in the policy for every single update. Occasionally, when they want to disallow a problematic version, it would be awkward to list all currently running/upgradable versions as allowed.

**Review by mprahl — APPROVED** [review]
/hold for other reviews

**Pull request merged**
